File: ruby/sql_injection/order_queries.rb
require 'active_record'
require 'pg'

# VULN 1: ActiveRecord string interpolation - order history
def get_order_history(user_id, status)
  Order.where("user_id = #{user_id} AND status = '#{status}'")
end

# VULN 2: String concatenation in pg query - orders by date range
def get_orders_by_date_range(status, start_date, end_date)
  conn = PG.connect(host: 'localhost', dbname: 'shopist')
  conn.exec(
    "SELECT * FROM orders WHERE status = '" + status +
    "' AND created_at BETWEEN '" + start_date + "' AND '" + end_date + "'"
  )
end

# VULN 3: find_by_sql with interpolation and JOIN - invoice lookup
def get_invoice_data(order_id, customer_name)
  Order.find_by_sql(
    "SELECT o.*, u.name, u.email FROM orders o " \
    "JOIN users u ON o.user_id = u.id " \
    "WHERE o.id = #{order_id} AND u.name = '#{customer_name}'"
  ).first
end

File: python/sql_injection/order_queries.py
import sqlite3
import psycopg2


# VULN 1: String concatenation SQL injection - order history
def get_order_history(user_id, status):
    conn = sqlite3.connect("shopist.db")
    cursor = conn.cursor()
    query = "SELECT * FROM orders WHERE user_id = " + str(user_id) + " AND status = '" + status + "'"
    cursor.execute(query)
    return cursor.fetchall()


# VULN 2: f-string SQL injection - orders by date range
def get_orders_by_date_range(status, start_date, end_date):
    conn = psycopg2.connect(host="db", dbname="shopist", user="app", password="apppass")
    cursor = conn.cursor()
    query = f"SELECT id, user_id, total, status FROM orders WHERE status = '{status}' AND created_at BETWEEN '{start_date}' AND '{end_date}'"
    cursor.execute(query)
    return cursor.fetchall()


# VULN 3: %-format SQL injection with JOIN - invoice lookup
def get_invoice_data(order_id, customer_name):
    conn = sqlite3.connect("shopist.db")
    cursor = conn.cursor()
    query = "SELECT o.*, u.name, u.email FROM orders o JOIN users u ON o.user_id = u.id WHERE o.id = %s AND u.name = '%s'" % (order_id, customer_name)
    cursor.execute(query)
    return cursor.fetchone()
